File: app/src/main/java/com/example/deltatask3/adapters/PokemonAdapter.java
package com.example.deltatask3.adapters;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.deltatask3.R;
import com.example.deltatask3.databinding.PokemonRowBinding;
import com.example.deltatask3.utils.Pokemon;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

public class PokemonAdapter extends RecyclerView.Adapter<PokemonAdapter.PokemonViewHolder> {

    private ArrayList<Pokemon> pokemons;
    private PokemonAdapterListener listener;

    public void setPokemons(ArrayList<Pokemon> pokemons) {
        this.pokemons = pokemons;
    }

    public void setListener(PokemonAdapterListener listener) {
        this.listener = listener;
    }

    public interface PokemonAdapterListener {
        void onItemClicked(int position, ImageView pokemon, TextView name);
    }

    public static class PokemonViewHolder extends RecyclerView.ViewHolder {

        PokemonRowBinding binding;

        public PokemonViewHolder(@NonNull PokemonRowBinding pokemonRowBinding, PokemonAdapterListener listener) {
            super(pokemonRowBinding.getRoot());
            binding = pokemonRowBinding;
            binding.row.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onItemClicked(getAdapterPosition(), binding.ivPokemon,binding.tvName);
                }
            });
        }
    }

    @NonNull
    @Override
    public PokemonViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        PokemonRowBinding binding = PokemonRowBinding.inflate(layoutInflater, parent, false);
        return new PokemonViewHolder(binding, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull PokemonViewHolder holder, int position) {
        Pokemon pokemon = pokemons.get(position);

        if (pokemon.getSprites() != null) {
            String urlSprite = pokemon.getSprites().getFront_default();
            if (urlSprite!=null && !urlSprite.isEmpty())
                Picasso.get().load(urlSprite).placeholder(R.drawable.placeholder_image).into(holder.binding.ivPokemon);
        }
        holder.binding.tvId.setText(String.valueOf(pokemon.getId()));
        holder.binding.tvName.setText(firstLetterToUppercase(pokemon.getName()));
    }

    @Override
    public int getItemCount() {
        return pokemons.size();
    }

    private String firstLetterToUppercase(String string){
        return string.substring(0,1).toUpperCase()+string.substring(1);
    }

}

File: app/src/main/java/com/example/deltatask3/adapters/PokemonAdapter.kt
package com.example.deltatask3.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.deltatask3.R
import com.example.deltatask3.adapters.PokemonAdapter.PokemonViewHolder
import com.example.deltatask3.databinding.PokemonRowBinding
import com.example.deltatask3.firstLetterToUppercase
import com.example.deltatask3.models.Pokemon
import com.squareup.picasso.Picasso
import java.util.*

class PokemonAdapter(
        var pokemons: ArrayList<Pokemon>,
        private var listener: PokemonAdapterListener
) : RecyclerView.Adapter<PokemonViewHolder>() {

    fun interface PokemonAdapterListener {
        fun onItemClicked(position: Int, pokemon: ImageView, name: TextView)
    }

    class PokemonViewHolder(var binding: PokemonRowBinding, listener: PokemonAdapterListener?) : RecyclerView.ViewHolder(binding.root) {
        init {
            binding.row.setOnClickListener { listener!!.onItemClicked(adapterPosition, binding.ivPokemon, binding.tvName) }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PokemonViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding = PokemonRowBinding.inflate(layoutInflater, parent, false)
        return PokemonViewHolder(binding, listener)
    }

    override fun onBindViewHolder(holder: PokemonViewHolder, position: Int) {
        val pokemon = pokemons[position]
        if (pokemon.sprites != null) {
            val urlSprite = pokemon.sprites.front_default
            if (urlSprite != null && urlSprite.isNotEmpty())
                Picasso.get().load(urlSprite).placeholder(R.drawable.placeholder_image).into(holder.binding.ivPokemon)
        }
        holder.binding.tvId.text = pokemon.id.toString()
        holder.binding.tvName.text = firstLetterToUppercase(pokemon.name)
    }

    override fun getItemCount() = pokemons.size

    fun getPokemonAt(pos: Int) = pokemons[pos]
}